**Analyze the topics**
--


We have collected the topics with the the help of the two sentence transformers: "allenai-specter" and "all-MiniLM-L6-v2". The former one is for academic purposes while the latter one is for the short texts. 

In [36]:
import pandas as pd
df_allenai = pd.read_csv("analyzed_data_new_allenai.csv")
df_allenai = df_allenai.drop('Unnamed: 0', axis=1)
topics_allenai = pd.read_csv("youtube-topics-new-allenai.csv")

In [38]:
df_minilm = pd.read_csv("analyzed_data_new_MiniLm.csv")
topics_minilm = pd.read_csv("youtube-topics-new-MiniLm.csv")

In [58]:
topics_allenai

,Topic,Count,Name,Representation,Representative_Docs
0,-1,9635,-1_biohacking_facial_aging_skin,"['biohacking', 'facial', 'aging', 'skin', 'hea...","['LCN Anti-aging Manicure ', ""Maxine Warsh Fac..."
1,0,2652,0_healthy_healthy aging_longevity_aging,"['healthy', 'healthy aging', 'longevity', 'agi...",['Int: Wise Healthy Aging1 Wise and Healthy Ag...
2,1,844,1_skincare_routine_skincare routine_serum,"['skincare', 'routine', 'skincare routine', 's...",['Anti Aging Skincare Routine For Women Over 5...
3,2,600,2_nutrition_foods_diet_aging foods,"['nutrition', 'foods', 'diet', 'aging foods', ...","[""The Longevity Diet - A Full Day Of Eating In..."
4,3,587,3_biohacking_el_en_que,"['biohacking', 'el', 'en', 'que', 'biohacker',...","['Biohacking y sus beneficios para la salud, e..."
...,...,...,...,...,...
178,177,11,177_laser_dual_fraxel_frank,"['laser', 'dual', 'fraxel', 'frank', 'med', 's...",['Anti-Aging Fraxel Dual Laser A Fall Routine ...
179,178,11,178_business_innovation_award_success,"['business', 'innovation', 'award', 'success',...","[""OCBJ Family Owned Business Awards- Longevity..."
180,179,10,179_excerpt_self_children_social,"['excerpt', 'self', 'children', 'social', 'ric...","[""Biohacking, Psychedelics, &amp; Physical Mas..."
181,180,10,180_technology_safe_supports_ways,"['technology', 'safe', 'supports', 'ways', 'se...",['Technology &amp; Healthy Aging: If It Were O...


In [60]:
topics_minilm

,Topic,Count,Name,Representation,Representative_Docs
0,-1,8459,-1_aging_longevity_healthy_nootropics,"['aging', 'longevity', 'healthy', 'nootropics'...",['How To Choose The Right Anti-Aging Cream Ant...
1,0,11355,0_rejuvenation_aging_anti_anti aging,"['rejuvenation', 'aging', 'anti', 'anti aging'...",['Nootropics for Hangovers: High Level Strateg...
2,1,484,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'...","['Biohacking y sus beneficios para la salud, e..."
3,2,185,2_rejuvenation_rejuvenation episode_game_play,"['rejuvenation', 'rejuvenation episode', 'game...",['Let&#39;s Play Pokemon Rejuvenation: Episode...
4,3,91,3_klaus_workout_denmark_2012,"['klaus', 'workout', 'denmark', '2012', '2011'...",['MUTANT athlete KLAUS RIIS will be at FIBO 20...
5,4,78,4_exercise video_ball_demonstrating_exercise,"['exercise video', 'ball', 'demonstrating', 'e...",['REJUVENATION | DELTOID FLIES Rejuvenation Pr...
6,5,76,5_sleep_melatonin_night_issues,"['sleep', 'melatonin', 'night', 'issues', 'hea...","['Yeah Yeah Yeah, Sleep... ""I\'ll Sleep When I..."
7,6,71,6_plasma_cutter_stick_140,"['plasma', 'cutter', 'stick', '140', 'longevit...",['LONGEVITY WELDALL TIG WELDER PLASMA CUTTER C...
8,7,70,7_vaginal_vaginal rejuvenation_rejuvenation_tr...,"['vaginal', 'vaginal rejuvenation', 'rejuvenat...","['⭕️ Vaginal Rejuvenation (Vaginoplasty, Vagin..."
9,8,62,8_record_mix_original_guide,"['record', 'mix', 'original', 'guide', 'album'...","[""1992 - Nootropic &#39;Elevation&#39; (Origin..."


***Start analyzing***
-

In [68]:
#for the later convenience
dfa   = df_allenai.copy()     
dfm   = df_minilm.copy()       
tpa   = topics_allenai.copy()  
tpm   = topics_minilm.copy()

In [70]:
# Keep only what's needed from the topics tables
tpa_small = tpa[['Topic', 'Name', 'Representation']].rename(
    columns={'Topic':'topic', 'Name':'topic_name', 'Representation':'topic_repr'}
)
tpm_small = tpm[['Topic', 'Name', 'Representation']].rename(
    columns={'Topic':'topic', 'Name':'topic_name', 'Representation':'topic_repr'}
)

# Left-join so every video row gets the topic label/keywords
dfa = dfa.merge(tpa_small, on='topic', how='left')
dfm = dfm.merge(tpm_small, on='topic', how='left')

In [74]:
def topic_diagnostics(df, topics_df, label):
    n_docs = len(df)
    n_topics = (topics_df['Topic'] != -1).sum()
    outliers = (df['topic'] == -1).sum()
    print(f"[{label}] docs={n_docs:,} | topics={n_topics} | outliers={outliers} ({outliers/n_docs:.2%})")
    display(topics_df.head())

topic_diagnostics(dfa, tpa, "SPECTER")
topic_diagnostics(dfm, tpm, "MiniLM")

[SPECTER] docs=22,295 | topics=182 | outliers=9635 (43.22%)


,Topic,Count,Name,Representation,Representative_Docs
0,-1,9635,-1_biohacking_facial_aging_skin,"['biohacking', 'facial', 'aging', 'skin', 'hea...","['LCN Anti-aging Manicure ', ""Maxine Warsh Fac..."
1,0,2652,0_healthy_healthy aging_longevity_aging,"['healthy', 'healthy aging', 'longevity', 'agi...",['Int: Wise Healthy Aging1 Wise and Healthy Ag...
2,1,844,1_skincare_routine_skincare routine_serum,"['skincare', 'routine', 'skincare routine', 's...",['Anti Aging Skincare Routine For Women Over 5...
3,2,600,2_nutrition_foods_diet_aging foods,"['nutrition', 'foods', 'diet', 'aging foods', ...","[""The Longevity Diet - A Full Day Of Eating In..."
4,3,587,3_biohacking_el_en_que,"['biohacking', 'el', 'en', 'que', 'biohacker',...","['Biohacking y sus beneficios para la salud, e..."


[MiniLM] docs=22,295 | topics=58 | outliers=8459 (37.94%)


,Topic,Count,Name,Representation,Representative_Docs
0,-1,8459,-1_aging_longevity_healthy_nootropics,"['aging', 'longevity', 'healthy', 'nootropics'...",['How To Choose The Right Anti-Aging Cream Ant...
1,0,11355,0_rejuvenation_aging_anti_anti aging,"['rejuvenation', 'aging', 'anti', 'anti aging'...",['Nootropics for Hangovers: High Level Strateg...
2,1,484,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'...","['Biohacking y sus beneficios para la salud, e..."
3,2,185,2_rejuvenation_rejuvenation episode_game_play,"['rejuvenation', 'rejuvenation episode', 'game...",['Let&#39;s Play Pokemon Rejuvenation: Episode...
4,3,91,3_klaus_workout_denmark_2012,"['klaus', 'workout', 'denmark', '2012', '2011'...",['MUTANT athlete KLAUS RIIS will be at FIBO 20...


From the above results, it seems like MiniLm worked better to our dataset since there are fewer topics and also 37.94% was considered as noisy. Let's analyze it first 

Let's seach for the unrelated topics and delete them. Our strategy will be deleting the unrelated topics and videos and re-run the Bert Topic once again. 

In [81]:
dfm[dfm['topic']==1].head(50)

,source,video_id,video_url,title,description,channel,comments,views,likes,comments_count,...,country,date,likes_missing,year,engagement,engagement_rate,text,topic,topic_name,topic_repr
40,YouTube,aiq8G9Gx5qo,https://www.youtube.com/watch?v=aiq8G9Gx5qo,Pontus ströbaek.m4v,NaN,Den Ökända Salong Betong AB,NaN,1235.0,2.0,0.0,...,unknown,2010-01-15,False,2010,2.0,0.001619,Pontus ströbaek.m4v,1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."
209,YouTube,U3EXfSecZ94,https://www.youtube.com/watch?v=U3EXfSecZ94,greenwheels.mpg,promofilmpje van electrice auto van green wheels.,Hans Steeman,NaN,2522.0,1.0,0.0,...,unknown,2010-03-08,False,2010,1.0,0.000397,greenwheels.mpg promofilmpje van electrice aut...,1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."
364,YouTube,5tnB2fBI0GQ,https://www.youtube.com/watch?v=5tnB2fBI0GQ,Mikocionis Deividas.wmv,Mikocionis Deividas.,ljowkabelov,"Blemba, Maladec ... Perspectyvus Tai sakes :} ...",738.0,4.0,1.0,...,unknown,2010-04-26,False,2010,5.0,0.006775,Mikocionis Deividas.wmv Mikocionis Deividas.,1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."
418,YouTube,Y-wRk0us1WY,https://www.youtube.com/watch?v=Y-wRk0us1WY,Gumiarc.3gp,hat ez kurva jo!!!=))=)),iissttii1,NaN,390.0,0.0,0.0,...,unknown,2010-05-10,False,2010,0.0,0.000000,Gumiarc.3gp hat ez kurva jo!!!=))=)),1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."
578,YouTube,hTPVY6ev-pY,https://www.youtube.com/watch?v=hTPVY6ev-pY,CPCO3 - BioHacking (1/2),Esta conferencia trabaja sobre los conceptos y...,Campus Party,NaN,3429.0,15.0,0.0,...,unknown,2010-06-30,False,2010,15.0,0.004374,CPCO3 - BioHacking (1/2) Esta conferencia trab...,1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."
991,YouTube,34aLp0zj7MQ,https://www.youtube.com/watch?v=34aLp0zj7MQ,Empoisonnement au plomb dans des mines d&#39;o...,"""Plus de 400 enfants"" sont morts d'empoisonnem...",AFP,NaN,450.0,2.0,0.0,...,unknown,2010-10-05,False,2010,2.0,0.004444,Empoisonnement au plomb dans des mines d&#39;o...,1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."
1169,YouTube,S7fpZExGF_g,https://www.youtube.com/watch?v=S7fpZExGF_g,Hirndoping - Auswirkungen der Umkehrung von mo...,http://konsumpf.de/?p=6743 Durch die Veränderu...,popscenede,Vollkommen frei erfundene Geschichte mit märch...,13801.0,34.0,5.0,...,unknown,2010-11-07,False,2010,39.0,0.002826,Hirndoping - Auswirkungen der Umkehrung von mo...,1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."
1174,YouTube,GzwRZja9JcM,https://www.youtube.com/watch?v=GzwRZja9JcM,Hirndoping - Auswirkungen der Umkehrung von mo...,http://konsumpf.de/?p=6743 Durch die Veränderu...,popscenede,Modafinil hat nichts mit den Amphetaminen zu t...,5625.0,10.0,3.0,...,unknown,2010-11-07,False,2010,13.0,0.002311,Hirndoping - Auswirkungen der Umkehrung von mo...,1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."
1426,YouTube,OFYLhuFxHRo,https://www.youtube.com/watch?v=OFYLhuFxHRo,Vivre avec le VIH,Vivre avec le VIH Pulsations TV - 103 - décemb...,Hopitaux Universitaires de Genève,Merci Seigneur 🙏 de m&#39;avoir conduit au Dr ...,25258.0,49.0,10.0,...,unknown,2011-01-03,False,2011,59.0,0.002336,Vivre avec le VIH Vivre avec le VIH Pulsations...,1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."
1527,YouTube,Yd01DRZV_NM,https://www.youtube.com/watch?v=Yd01DRZV_NM,Lecenje od heroinske zavisnosti.Zavisnost od o...,Posle 22 godine zloupotrebe raznih vrsta narko...,Naltrexone implants,NaN,1441.0,0.0,0.0,...,unknown,2011-01-26,True,2011,0.0,0.000000,Lecenje od heroinske zavisnosti.Zavisnost od o...,1,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'..."


As we can see from above, the topic number 1 of the Minilm consists of spanish, germand and portugeese videos. Those videos are not needed in our research, since it is based on English langugage. So deleting them will not hurt

**Search for the other unrelated topics**

In [91]:
search_keywords = ["longevity", "healthy aging", "biohacking", "rejuvenation", "anti-aging", "nootropics"]

In [95]:
def topic_is_relevant(rep, search_keywords):
    topic_words = eval(rep)  # convert string list to real list
    topic_words = [word.lower() for word in topic_words]
    return any(kw in word for word in topic_words for kw in search_keywords)

# Apply this to all topics
tpm['is_relevant'] = tpm['Representation'].apply(lambda rep: topic_is_relevant(rep, search_keywords))

In [143]:
tpm[tpm['is_relevant']==False]

,Topic,Count,Name,Representation,Representative_Docs,is_relevant,relevance_ratio
4,3,91,3_klaus_workout_denmark_2012,"['klaus', 'workout', 'denmark', '2012', '2011'...",['MUTANT athlete KLAUS RIIS will be at FIBO 20...,False,0.000000
5,4,78,4_exercise video_ball_demonstrating_exercise,"['exercise video', 'ball', 'demonstrating', 'e...",['REJUVENATION | DELTOID FLIES Rejuvenation Pr...,False,0.538462
9,8,62,8_record_mix_original_guide,"['record', 'mix', 'original', 'guide', 'album'...","[""1992 - Nootropic &#39;Elevation&#39; (Origin...",False,0.000000
10,9,60,9_bulletproof_asprey_bulletproof coffee_dave a...,"['bulletproof', 'asprey', 'bulletproof coffee'...",['Bulletproof Exec Dave Asprey on BioHacking a...,False,0.483333
11,10,60,10_на_39 play_let 39_nzt,"['на', '39 play', 'let 39', 'nzt', 'hd', 'play...",['ЭФФЕКТИВНЫЕ упражнения для омоложения! #омол...,False,0.100000
13,12,47,12_arduino_shield_game_games,"['arduino', 'shield', 'game', 'games', 'video'...",['Video Overlay Goggles Arduino Using an Ardui...,False,0.021277
14,13,47,13_oral_dental_oral health_dentistry,"['oral', 'dental', 'oral health', 'dentistry',...",['Nutrition and Oral Health | Aging Matters | ...,False,0.574468
15,14,46,14_limitless_cooper_nzt_movie,"['limitless', 'cooper', 'nzt', 'movie', 'trail...",['Limitless Movie Review from Haunted Flower G...,False,0.043478
16,15,45,15_diy bio_diy_bio_pellet,"['diy bio', 'diy', 'bio', 'pellet', 'water', '...","['DIY bio pellets reactor 1 ', 'DIY Bio-Pellet...",False,0.088889
18,17,37,17_lower_song_brains_click http,"['lower', 'song', 'brains', 'click http', 'pla...","[""Lower Dens - Propagation After just 76 secon...",False,0.675676


As you can see, I have filtered out unrelated topics based on the keywords. We will look at each cluster and decide if it should be deleted or left. In each topic, we will check if there might be related videos and how percentage are related. Based on that we will also manually look at them. If the topic is completely unrelated, then we will delete them.  

In [120]:
#function to check if the video is related or unrelated

def is_video_related(text, keywords):
    text = str(text).lower()
    return any(kw in text for kw in keywords)

In [122]:
#function to check how many percentage of the videos in the topics are related
def check_relevance_ratio(df, topic_id, keywords):
    df_topic = df[df['topic']==topic_id]
    relevant_count = df_topic['text'].apply(lambda x: is_video_related(x, keywords)).sum()
    return relevant_count/len(df_topic)

In [124]:
keywords = ["longevity", "healthy aging", "biohacking", "rejuvenation", "anti-aging", "nootropics"]

In [132]:
check_relevance_ratio(dfm, 4, keywords)

0.5384615384615384

Let's assign their relevance ratio for each topic

In [135]:
tpm['relevance_ratio'] = tpm['Topic'].apply(lambda t: check_relevance_ratio(dfm, t, keywords)) 

Now, all the topics have their own relevance ratio. Let's check the topics of which relecane ratio is smaller than .20 and if the videos are totally unrelated, we will delete them. 

In [149]:
tpm[tpm['relevance_ratio']<=0.20]

,Topic,Count,Name,Representation,Representative_Docs,is_relevant,relevance_ratio
4,3,91,3_klaus_workout_denmark_2012,"['klaus', 'workout', 'denmark', '2012', '2011'...",['MUTANT athlete KLAUS RIIS will be at FIBO 20...,False,0.000000
9,8,62,8_record_mix_original_guide,"['record', 'mix', 'original', 'guide', 'album'...","[""1992 - Nootropic &#39;Elevation&#39; (Origin...",False,0.000000
11,10,60,10_на_39 play_let 39_nzt,"['на', '39 play', 'let 39', 'nzt', 'hd', 'play...",['ЭФФЕКТИВНЫЕ упражнения для омоложения! #омол...,False,0.100000
13,12,47,12_arduino_shield_game_games,"['arduino', 'shield', 'game', 'games', 'video'...",['Video Overlay Goggles Arduino Using an Ardui...,False,0.021277
15,14,46,14_limitless_cooper_nzt_movie,"['limitless', 'cooper', 'nzt', 'movie', 'trail...",['Limitless Movie Review from Haunted Flower G...,False,0.043478
16,15,45,15_diy bio_diy_bio_pellet,"['diy bio', 'diy', 'bio', 'pellet', 'water', '...","['DIY bio pellets reactor 1 ', 'DIY Bio-Pellet...",False,0.088889
20,19,36,19_neurofeedback_brain_therapy_http,"['neurofeedback', 'brain', 'therapy', 'http', ...",['Is Neurofeedback Therapy a Useful Therapy? I...,False,0.083333
21,20,36,20_playlists_com ref_ref_visit https,"['playlists', 'com ref', 'ref', 'visit https',...",['Jiu Jitsu Self-Defense vs Weapons - How to D...,False,0.027778
24,23,29,23_smart drugs_drugs_smart_le,"['smart drugs', 'drugs', 'smart', 'le', 'di', ...",['Quanto costano le smart drugs? LOMBARDIA CHA...,False,0.034483
30,29,26,29_stave age_stave_drugs stave_develop smart,"['stave age', 'stave', 'drugs stave', 'develop...",['llive discussion can we should we devel...,False,0.000000


In [340]:
# id's of the topics that should be deleted
# 1? 
# 10 is related, but all the videos are in Russian
# 23 in Italian
# 31 is the videos of the Bulletproof channel. It is moslty related to the topic
# 47 is mostly about caring for the elder people. Not reletated to our research
delete_id =[3, 8, 10, 12, 14, 15, 20, 23, 34, 39, 47, 55]

Let's check out the vides between .20 and .50

In [208]:
tpm[(tpm['relevance_ratio']>=0.20) & (tpm['relevance_ratio']<=0.50)]

,Topic,Count,Name,Representation,Representative_Docs,is_relevant,relevance_ratio
10,9,60,9_bulletproof_asprey_bulletproof coffee_dave a...,"['bulletproof', 'asprey', 'bulletproof coffee'...",['Bulletproof Exec Dave Asprey on BioHacking a...,False,0.483333
22,21,33,21_alpha brain_alpha_onnit_brain,"['alpha brain', 'alpha', 'onnit', 'brain', 'ro...",['Joey Diaz gives his thoughts on Alpha Brain ...,False,0.242424
26,25,28,25_choline_alpha gpc_gpc_alpha,"['choline', 'alpha gpc', 'gpc', 'alpha', 'comp...",['Alpha-GPC - Nootropics Information What is A...,False,0.464286
32,31,25,31_bulletproof radio_radio_bulletproof_podcast,"['bulletproof radio', 'radio', 'bulletproof', ...",['Bulletproof Radio Q&amp;A - #207 Why you sho...,False,0.200000
43,42,20,42_piel_la_cómo_las,"['piel', 'la', 'cómo', 'las', 'del', 'el', 'qu...",['💆🏻\u200d♀️¿Cómo mejorar la texturade la piel...,False,0.250000
46,45,19,45_creatine_muscle_power_strength,"['creatine', 'muscle', 'power', 'strength', 'b...","[""Creatine &amp; Healthy Aging: Myth vs. Reali...",False,0.368421
49,48,18,48_silicon_complete_implants_extension,"['silicon', 'complete', 'implants', 'extension...",['27C3 Cybernetics for the Masses - Lepht Anon...,False,0.277778
50,49,18,49_magnet_implant_clients_stuff,"['magnet', 'implant', 'clients', 'stuff', 'sho...",['Magnet &amp; Paperclips #5 30 paperclips on ...,False,0.333333
52,51,18,51_2600 hope_2600_hope 2008_2008,"['2600 hope', '2600', 'hope 2008', '2008', 'ho...",['2600 - The Last HOPE (2008) - Bagcam_02 2600...,True,0.277778
54,53,17,53_stress_relief_professionals_teach,"['stress', 'relief', 'professionals', 'teach',...",['Stress Relief Anti Aging Tip Your Top 3 Prio...,False,0.470588


In [342]:
# 9 and 31 is about the Dave asprey videos, The founder of the Bulletproof company which focuses on the human health
# 42 all the vidoes are in Spanish
# 48 - cybernetics ? 
delete_id_2 =[42, 48, 51]

Let's now check the videos between 50 and 75 relevance

In [233]:
tpm[(tpm['relevance_ratio']>0.50) & (tpm['relevance_ratio']<=0.75)]

,Topic,Count,Name,Representation,Representative_Docs,is_relevant,relevance_ratio
0,-1,8459,-1_aging_longevity_healthy_nootropics,"['aging', 'longevity', 'healthy', 'nootropics'...",['How To Choose The Right Anti-Aging Cream Ant...,True,0.705521
5,4,78,4_exercise video_ball_demonstrating_exercise,"['exercise video', 'ball', 'demonstrating', 'e...",['REJUVENATION | DELTOID FLIES Rejuvenation Pr...,False,0.538462
6,5,76,5_sleep_melatonin_night_issues,"['sleep', 'melatonin', 'night', 'issues', 'hea...","['Yeah Yeah Yeah, Sleep... ""I\'ll Sleep When I...",True,0.513158
14,13,47,13_oral_dental_oral health_dentistry,"['oral', 'dental', 'oral health', 'dentistry',...",['Nutrition and Oral Health | Aging Matters | ...,False,0.574468
18,17,37,17_lower_song_brains_click http,"['lower', 'song', 'brains', 'click http', 'pla...","[""Lower Dens - Propagation After just 76 secon...",False,0.675676
29,28,27,28_gut_microbiome_bacteria_gut health,"['gut', 'microbiome', 'bacteria', 'gut health'...",['Gut Microbes &amp; Longevity: The Secret to ...,False,0.629630
45,44,19,44_theanine_caffeine_review_benefits,"['theanine', 'caffeine', 'review', 'benefits',...",['Top 4 L Theanine Benefits ltheanine #biohack...,True,0.526316
47,46,19,46_cold_ice_water_actually,"['cold', 'ice', 'water', 'actually', 'usa', 'b...",['Beyond Cold Water Boot Camp USA - Dive Door ...,True,0.578947
53,52,17,52_hearing_loss_public health_healthy,"['hearing', 'loss', 'public health', 'healthy'...",['CLSA Webinar Series: Hearing loss and health...,True,0.647059


In [344]:
# let's explore the -1 at the end since it is a noise
# 17 mostly songs and remix which are not related

delete_id_3 = [17]

Let's now check the videos between 0.75 and 100 relevance

In [254]:
tpm[(tpm['relevance_ratio']>0.75) & (tpm['relevance_ratio']<=1)]

,Topic,Count,Name,Representation,Representative_Docs,is_relevant,relevance_ratio
1,0,11355,0_rejuvenation_aging_anti_anti aging,"['rejuvenation', 'aging', 'anti', 'anti aging'...",['Nootropics for Hangovers: High Level Strateg...,True,0.805020
2,1,484,1_biohacking_la_en_el,"['biohacking', 'la', 'en', 'el', 'para', 'que'...","['Biohacking y sus beneficios para la salud, e...",True,0.869835
3,2,185,2_rejuvenation_rejuvenation episode_game_play,"['rejuvenation', 'rejuvenation episode', 'game...",['Let&#39;s Play Pokemon Rejuvenation: Episode...,True,0.972973
7,6,71,6_plasma_cutter_stick_140,"['plasma', 'cutter', 'stick', '140', 'longevit...",['LONGEVITY WELDALL TIG WELDER PLASMA CUTTER C...,True,0.985915
8,7,70,7_vaginal_vaginal rejuvenation_rejuvenation_tr...,"['vaginal', 'vaginal rejuvenation', 'rejuvenat...","['⭕️ Vaginal Rejuvenation (Vaginoplasty, Vagin...",True,0.942857
12,11,56,11_ovarian_fertility_prp_platelet rich,"['ovarian', 'fertility', 'prp', 'platelet rich...",['Ovarian PRP (Platelet-Rich Plasma)|Ovarian R...,True,0.964286
17,16,41,16_microneedling_exosomes_micro_skin,"['microneedling', 'exosomes', 'micro', 'skin',...",['How does MICRONEEDLING reverse signs of agin...,True,0.878049
19,18,37,18_aubrey_grey_aubrey grey_escape velocity,"['aubrey', 'grey', 'aubrey grey', 'escape velo...",['Aubrey de Grey | Longevity Myth Busting Alli...,True,0.918919
23,22,30,22_nootropics_fuel_motivation_edge,"['nootropics', 'fuel', 'motivation', 'edge', '...",['Korvyx - No Fairy Dust #brainupgrade #nootro...,True,0.966667
25,24,29,24_rich plasma_platelet rich_platelet_prp,"['rich plasma', 'platelet rich', 'platelet', '...",['PRP Hair Restoration - Hair Rejuvenation and...,True,0.931034


In [346]:
# 1 since al the videos are in different language
# 2 mostly about the pokemon rejuvenation game
# 6 mostly about the longevity welding
# 27 short vidoes of the longevity USA company event. Since it is about the event itself, not so much correlated with our topic
# 37 also different language
# 43 it disscusses the longevity of teh dogs and pets
# 54 in german

delete_id_4 = [1, 2, 6, 27, 33, 37, 54, 56, 57]

In [336]:
dfm[dfm['topic']==57].head(50)

,source,video_id,video_url,title,description,channel,comments,views,likes,comments_count,...,country,date,likes_missing,year,engagement,engagement_rate,text,topic,topic_name,topic_repr
1557,YouTube,-6Vx_-mltWg,https://www.youtube.com/watch?v=-6Vx_-mltWg,Welcome to The Longevity Garden,http://TheLongevityGarden.com -- Welcome to Th...,TheLongevityGarden,NaN,1380.0,3.0,0.0,...,unknown,2011-02-03,False,2011,3.0,0.002174,Welcome to The Longevity Garden http://TheLong...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."
1558,YouTube,Rnl9aoQVVf4,https://www.youtube.com/watch?v=Rnl9aoQVVf4,Longevity Garden Welcome.mp4,http://bit.ly/gZyjN5 Come playfully embrace lo...,LongevityGarden,NaN,15.0,0.0,0.0,...,unknown,2011-02-03,False,2011,0.0,0.000000,Longevity Garden Welcome.mp4 http://bit.ly/gZy...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."
1560,YouTube,JWeeMM5WnLs,https://www.youtube.com/watch?v=JWeeMM5WnLs,Longevity Garden Benefits.mp4,http://bit.ly/gZyjN5 Learn the Benefits of the...,LongevityGarden,NaN,7.0,0.0,0.0,...,unknown,2011-02-03,False,2011,0.0,0.000000,Longevity Garden Benefits.mp4 http://bit.ly/gZ...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."
1561,YouTube,iPo4RxPgDcA,https://www.youtube.com/watch?v=iPo4RxPgDcA,The Benefits of The Longevity Garden,http://TheLongevityGarden.com -- Welcome to th...,TheLongevityGarden,NaN,10.0,0.0,0.0,...,unknown,2011-02-03,False,2011,0.0,0.000000,The Benefits of The Longevity Garden http://Th...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."
1562,YouTube,bNG70mwvhUw,https://www.youtube.com/watch?v=bNG70mwvhUw,Longevity Garden Description.mp4,http://bit.ly/gZyjN5 Step in as Longevity Expe...,LongevityGarden,NaN,19.0,0.0,0.0,...,unknown,2011-02-03,False,2011,0.0,0.000000,Longevity Garden Description.mp4 http://bit.ly...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."
1564,YouTube,aU6JL5e4v6s,https://www.youtube.com/watch?v=aU6JL5e4v6s,Longevity Garden Fear Tree.mp4,http://bit.ly/gZyjN5 Of all the trees in our g...,LongevityGarden,NaN,26.0,0.0,0.0,...,unknown,2011-02-03,False,2011,0.0,0.000000,Longevity Garden Fear Tree.mp4 http://bit.ly/g...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."
1565,YouTube,uJC8E-rt92A,https://www.youtube.com/watch?v=uJC8E-rt92A,The Longevity Garden Description,http://TheLongevityGarden.com-- The Longevity ...,TheLongevityGarden,NaN,225.0,0.0,0.0,...,unknown,2011-02-03,False,2011,0.0,0.000000,The Longevity Garden Description http://TheLon...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."
1569,YouTube,DLxr_r_JMvI,https://www.youtube.com/watch?v=DLxr_r_JMvI,Longevity Garden Free Tree part 2.mp4,http://bit.ly/gZyjN5 Of all the trees in our g...,LongevityGarden,NaN,42.0,0.0,0.0,...,unknown,2011-02-06,False,2011,0.0,0.000000,Longevity Garden Free Tree part 2.mp4 http://b...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."
1591,YouTube,LOwf52LYUH4,https://www.youtube.com/watch?v=LOwf52LYUH4,The Longevity Garden Description,"http:/thelongevitygarden.com,Longevity, Alison...",TheLongevityGarden,NaN,29.0,0.0,0.0,...,unknown,2011-02-13,False,2011,0.0,0.000000,The Longevity Garden Description http:/thelong...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."
1596,YouTube,Q_YMYaQ8k8o,https://www.youtube.com/watch?v=Q_YMYaQ8k8o,Longevity Garden Attachment Tree.mp4,This tree of attachment has learned to let go ...,LongevityGarden,NaN,8.0,0.0,0.0,...,unknown,2011-02-14,False,2011,0.0,0.000000,Longevity Garden Attachment Tree.mp4 This tree...,57,57_garden_tree_mp4_trust,"['garden', 'tree', 'mp4', 'trust', 'descriptio..."


In [348]:
topics_to_delete = delete_id + delete_id_2 + delete_id_3 + delete_id_4
topics_to_delete

[3,
 8,
 10,
 12,
 14,
 15,
 20,
 23,
 34,
 39,
 47,
 55,
 42,
 48,
 51,
 17,
 1,
 2,
 6,
 27,
 33,
 37,
 54,
 56,
 57]

In [350]:
for topic_id in topics_to_delete:
    dfm = dfm[dfm['topic']!=topic_id]

In [367]:
mask = tpm['Topic'].apply(lambda x: x not in topics_to_delete)

In [369]:
tpm_new = tpm [mask]

**Let's save the final results and re-run the bertopic on the cleaned (from unrelated videos) dataset (Minilm SentenceTransformer)**

In [385]:
tpm_new.to_csv("Minilm-sorted-topics.csv", index=False)

In [387]:
dfm.to_csv("Minilm_sorted_data.csv", index=False)

**Let's run the bertopic once again on the cleaned dataset**

In [398]:
from bertopic import BERTopic
from sentence_transformers import SentenceTransformer
from sklearn.feature_extraction.text import CountVectorizer

model = SentenceTransformer("all-MiniLM-L6-v2") #for short texts

# Custom preprocessing with vectorizer
vectorizer_model = CountVectorizer(
    stop_words="english",  
    ngram_range=(1, 2),    
    max_df=0.85,           
    min_df=2,              
    max_features=4000  
)

# Creating BERTopic model
topic_model = BERTopic(
    embedding_model=model,
    vectorizer_model=vectorizer_model,
    nr_topics="auto",
    top_n_words=10,       
    calculate_probabilities=True,
    min_topic_size=30 #reduces overly small clusters
)

In [400]:
texts = dfm['text'].fillna("").tolist()

In [402]:
topics, probs = topic_model.fit_transform(texts)

In [406]:
topic_model.get_topic_info()

,Topic,Count,Name,Representation,Representative_Docs
0,-1,6947,-1_aging_nootropics_healthy_anti,"[aging, nootropics, healthy, anti, anti aging,...",[Anti-aging Skincare Secret - Part I The most ...
1,0,9492,0_rejuvenation_aging_longevity_anti,"[rejuvenation, aging, longevity, anti, anti ag...","[Rejuvenation , RevLite Skin Rejuvenation , Sk..."
2,1,2739,1_biohacking_biohacker_biohack_bulletproof,"[biohacking, biohacker, biohack, bulletproof, ...",[Biohacking Luisa Fernanda González Cárdenas Y...
3,2,996,2_nootropics_cortex_nootropic_stack,"[nootropics, cortex, nootropic, stack, nootrop...",[Nootropics make me agitated? Shorter Fuse? We...
4,3,212,3_drugs_smart_smart drugs_adderall,"[drugs, smart, smart drugs, adderall, students...","[Smart Drugs , Smart Drugs , Smart Drugs Diary ]"
5,4,75,4_anti_anti aging_aging_quot,"[anti, anti aging, aging, quot, doctor, talk, ...","[ศาสตร์มืด, แพง, เรื่องไกลตัว !? สรุปแล้ว Anti..."
6,5,69,5_coffee_mushroom_nootropics_mushrooms,"[coffee, mushroom, nootropics, mushrooms, savv...",[No More Coffee Runs! #Nootropics #Coffee #mid...
7,6,65,6_exercise video_video demonstrating_demonstra...,"[exercise video, video demonstrating, demonstr...",[REJUVENATION | SHOULDER PRESS Rejuvenation Pr...
8,7,55,7_ovarian_prp_fertility_rejuvenation,"[ovarian, prp, fertility, rejuvenation, ivf, p...",[My experience getting Ovarian PRP – Does it r...
9,8,43,8_oral_dental_dentistry_oral health,"[oral, dental, dentistry, oral health, implant...",[Nutrition and Oral Health | Aging Matters | N...


**Saving the results**

In [428]:
dfm['topic'] = topics

In [422]:
tpm_rerun = topic_model.get_topic_info()

In [424]:
dfm.drop(['topic_name', 'topic_repr'], axis=1, inplace=True)

In [432]:
tpm_small = tpm_rerun[['Topic', 'Name', 'Representation']].rename(
    columns={'Topic':'topic', 'Name':'topic_name', 'Representation':'topic_repr'}
)

# Left-join so every video row gets the topic label/keywords
dfm = dfm.merge(tpm_small, on='topic', how='left')

In [442]:
import numpy as np

dfm['topic_prob'] = [p[t] if t != -1 and p is not None else np.nan for p, t in zip(probs, topics)]

In [446]:
dfm.drop('topic_prob', axis=1, inplace=True)

In [450]:
dfm.to_csv("new_minilm_data.csv", index=False)

In [454]:
tpm_rerun.to_csv("new_minilm_topics.csv", index=False)

In [456]:
import pandas as pd

probs_df = pd.DataFrame(probs)
probs_df.columns = [f"Topic_{i}" for i in range(probs_df.shape[1])]

In [460]:
probs_df.to_csv("topic_probs_minilm.csv", index=False)

**Let's try another sentence transformer for the new data**

In [463]:
from bertopic import BERTopic
from sentence_transformers import SentenceTransformer
from sklearn.feature_extraction.text import CountVectorizer

model = SentenceTransformer("paraphrase-mpnet-base-v2") #more nuanced model 

# Custom preprocessing with vectorizer
vectorizer_model = CountVectorizer(
    stop_words="english",  
    ngram_range=(1, 2),    
    max_df=0.85,           
    min_df=2,              
    max_features=4000  
)

# Creating BERTopic model
topic_model = BERTopic(
    embedding_model=model,
    vectorizer_model=vectorizer_model,
    nr_topics="auto",
    top_n_words=10,       
    calculate_probabilities=True,
    min_topic_size=30 #reduces overly small clusters
)

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/594 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [465]:
topics, probs = topic_model.fit_transform(texts)

In [489]:
topic_model.get_topic_info().to_csv("new_paraphrase_topics.csv", index=False)

In [477]:
dfm.drop(['topic_name', 'topic_repr'], axis=1, inplace=True)

In [479]:
dfm['topic'] = topics

In [483]:
tpm_small = topic_model.get_topic_info()[['Topic', 'Name', 'Representation']].rename(
    columns={'Topic':'topic', 'Name':'topic_name', 'Representation':'topic_repr'}
)

# Left-join so every video row gets the topic label/keywords
dfm = dfm.merge(tpm_small, on='topic', how='left')

In [487]:
dfm.to_csv("new_paraphrase_dataset.csv", index=False)

Paraphrases MPnet SentenceTransformer analysis
-

Let's analyze the topics given by the Paraphrase Sentence Transformer. 

In [502]:
import pandas as pd

dfp = pd.read_csv("new_paraphrase_dataset.csv")
tpp = pd.read_csv("new_paraphrase_topics.csv")

In [504]:
tpp

,Topic,Count,Name,Representation,Representative_Docs
0,-1,6094,-1_longevity_healthy_healthy aging_anti,"['longevity', 'healthy', 'healthy aging', 'ant...",['The Healthy Ageing Secret The secret to heal...
1,0,3235,0_nootropics_nootropic_brain_stack,"['nootropics', 'nootropic', 'brain', 'stack', ...",['#lionsmane #lionsmanemushroom #nootropics #c...
2,1,2794,1_skin_anti_anti aging_facial,"['skin', 'anti', 'anti aging', 'facial', 'faci...","['Facial rejuvenation ', 'Dr Kusuma: What is ..."
3,2,2746,2_biohacking_biohacker_biohack_bulletproof,"['biohacking', 'biohacker', 'biohack', 'bullet...",['А вы следите за уровнем глюкозы? #биохакинг ...
4,3,1084,3_rejuvenation_music_provided youtube_provided,"['rejuvenation', 'music', 'provided youtube', ...",['Cultured Moments for Rejuvenation Provided t...
5,4,779,4_diet_foods_nutrition_longevity,"['diet', 'foods', 'nutrition', 'longevity', 'e...","[""We did the ProLon five day fast-mimicking lo..."
6,5,676,5_longevity_longevity tv_jeunesse_tv,"['longevity', 'longevity tv', 'jeunesse', 'tv'...",['Jeunesse Longevity TV PL Episode 1 1080P ...
7,6,642,6_healthy aging_healthy_health_healthy ageing,"['healthy aging', 'healthy', 'health', 'health...","['Healthy Aging Kin 250.', 'Healthy Aging RC v..."
8,7,563,7_exercise_fitness_yoga_exercises,"['exercise', 'fitness', 'yoga', 'exercises', '...",['Exercise for Health and Longevity Do you nee...
9,8,265,8_brain_alzheimer_dementia_aging brain,"['brain', 'alzheimer', 'dementia', 'aging brai...","[""Healthy Brain Aging &amp; Dementia Preventio..."


Let's check the relevance ratio of the each topic with the previosly applied method

In [509]:
#function to check if the video is related or unrelated

def is_video_related(text, keywords):
    text = str(text).lower()
    return any(kw in text for kw in keywords)

In [511]:
#function to check how many percentage of the videos in the topics are related
def check_relevance_ratio(df, topic_id, keywords):
    df_topic = df[df['topic']==topic_id]
    relevant_count = df_topic['text'].apply(lambda x: is_video_related(x, keywords)).sum()
    return relevant_count/len(df_topic)

keywords = ["longevity", "healthy aging", "biohacking", "rejuvenation", "anti-aging", "nootropics"]

In [513]:
tpp['relevance_ratio'] = tpp['Topic'].apply(lambda t: check_relevance_ratio(dfp, t, keywords)) 

In [515]:
tpp

,Topic,Count,Name,Representation,Representative_Docs,relevance_ratio
0,-1,6094,-1_longevity_healthy_healthy aging_anti,"['longevity', 'healthy', 'healthy aging', 'ant...",['The Healthy Ageing Secret The secret to heal...,0.778142
1,0,3235,0_nootropics_nootropic_brain_stack,"['nootropics', 'nootropic', 'brain', 'stack', ...",['#lionsmane #lionsmanemushroom #nootropics #c...,0.640185
2,1,2794,1_skin_anti_anti aging_facial,"['skin', 'anti', 'anti aging', 'facial', 'faci...","['Facial rejuvenation ', 'Dr Kusuma: What is ...",0.803865
3,2,2746,2_biohacking_biohacker_biohack_bulletproof,"['biohacking', 'biohacker', 'biohack', 'bullet...",['А вы следите за уровнем глюкозы? #биохакинг ...,0.691916
4,3,1084,3_rejuvenation_music_provided youtube_provided,"['rejuvenation', 'music', 'provided youtube', ...",['Cultured Moments for Rejuvenation Provided t...,0.882841
5,4,779,4_diet_foods_nutrition_longevity,"['diet', 'foods', 'nutrition', 'longevity', 'e...","[""We did the ProLon five day fast-mimicking lo...",0.820282
6,5,676,5_longevity_longevity tv_jeunesse_tv,"['longevity', 'longevity tv', 'jeunesse', 'tv'...",['Jeunesse Longevity TV PL Episode 1 1080P ...,0.980769
7,6,642,6_healthy aging_healthy_health_healthy ageing,"['healthy aging', 'healthy', 'health', 'health...","['Healthy Aging Kin 250.', 'Healthy Aging RC v...",0.691589
8,7,563,7_exercise_fitness_yoga_exercises,"['exercise', 'fitness', 'yoga', 'exercises', '...",['Exercise for Health and Longevity Do you nee...,0.824156
9,8,265,8_brain_alzheimer_dementia_aging brain,"['brain', 'alzheimer', 'dementia', 'aging brai...","[""Healthy Brain Aging &amp; Dementia Preventio...",0.422642


Relevance ratio of the all topics are more or less good. Let's analyze the topics number 11 and 12 since they seem like consisting of vidieo which are not english

In [524]:
dfp[dfp['topic']==11].head(50)

,source,video_id,video_url,title,description,channel,comments,views,likes,comments_count,...,country,date,likes_missing,year,engagement,engagement_rate,text,topic,topic_name,topic_repr
190,YouTube,p42oD2ycxi8,https://www.youtube.com/watch?v=p42oD2ycxi8,Chilena exigió medicamento para cáncer,Una mujer llegó a los tribunales exigiendo que...,Univision,NaN,261.0,0.0,0.0,...,unknown,2010-03-05,False,2010,0.0,0.000000,Chilena exigió medicamento para cáncer Una muj...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."
433,YouTube,y8p6yRq_62Y,https://www.youtube.com/watch?v=y8p6yRq_62Y,Pieles rejuvenecidas con las últimas técnicas ...,Vídeo para un evento deportivo con el que Clín...,Puntual Webmaster,NaN,84.0,0.0,0.0,...,unknown,2010-05-26,False,2010,0.0,0.000000,Pieles rejuvenecidas con las últimas técnicas ...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."
465,YouTube,x1l8Zho5rA4,https://www.youtube.com/watch?v=x1l8Zho5rA4,LaCore anti aging Promo Español,bHIP GLOBAL ha desarrollado la línea de produc...,sammuss,NaN,455.0,2.0,0.0,...,unknown,2010-06-04,False,2010,2.0,0.004396,LaCore anti aging Promo Español bHIP GLOBAL ha...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."
490,YouTube,1zHt5fx2O1M,https://www.youtube.com/watch?v=1zHt5fx2O1M,VITAMINE PER LA MENTE - LO SPECCHIO,Direttamente dal Format ''Sono un Genio'' del ...,Emiliano Schincaglia,NaN,138.0,0.0,0.0,...,unknown,2010-06-11,False,2010,0.0,0.000000,VITAMINE PER LA MENTE - LO SPECCHIO Direttamen...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."
640,YouTube,4HPLwj_WDNk,https://www.youtube.com/watch?v=4HPLwj_WDNk,Pharmavita RENACENZ TVN Enlaces parte2.mov,2º Parte de el programa Enlaces de TVN realiza...,Pharmavita,NaN,2706.0,12.0,0.0,...,unknown,2010-07-22,False,2010,12.0,0.004435,Pharmavita RENACENZ TVN Enlaces parte2.mov 2º ...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."
641,YouTube,enYjNAEq4Mc,https://www.youtube.com/watch?v=enYjNAEq4Mc,Renacenz testimonios reales,El noticiario Meganoticias realiza reportaje s...,Pharmavita,NaN,13005.0,50.0,0.0,...,unknown,2010-07-22,False,2010,50.0,0.003845,Renacenz testimonios reales El noticiario Mega...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."
649,YouTube,Yq75McpR02E,https://www.youtube.com/watch?v=Yq75McpR02E,Sei nuove droghe,22/07/10 (1536-69) - Sono state scoperte 6 nuo...,SteadyInfo,w la Marijuana terapeutica le altre sostanze s...,3114.0,5.0,5.0,...,unknown,2010-07-26,False,2010,10.0,0.003211,Sei nuove droghe 22/07/10 (1536-69) - Sono sta...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."
685,YouTube,Q74bPzxz9lM,https://www.youtube.com/watch?v=Q74bPzxz9lM,Questura di Verona - Sequestro smart-shop di V...,"JWH-018 e JWH-073, si presentano come profumi ...",Osservatorio dipendenze,NaN,2069.0,4.0,0.0,...,unknown,2010-08-04,False,2010,4.0,0.001933,Questura di Verona - Sequestro smart-shop di V...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."
871,YouTube,jL4jbDxw7tY,https://www.youtube.com/watch?v=jL4jbDxw7tY,Fashionista Recommends Anti-Aging Formula,1-800-260-9404 http://www.MAQUILLAJE.co entrev...,AntiAgingSkinCompany,NaN,103.0,1.0,0.0,...,unknown,2010-09-26,False,2010,1.0,0.009709,Fashionista Recommends Anti-Aging Formula 1-80...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."
895,YouTube,FLkjpUX63cM,https://www.youtube.com/watch?v=FLkjpUX63cM,Cremes para rugas,Os cremes antirrugas ou antissinais bloqueiam ...,Minha Vida,Ou morrendo cedo kkkkkkk || Só cirurgia plásti...,50276.0,278.0,4.0,...,unknown,2010-09-30,False,2010,282.0,0.005609,Cremes para rugas Os cremes antirrugas ou anti...,11,11_di_el_la_que,"['di', 'el', 'la', 'que', 'il', 'medicina', 'p..."


In [526]:
dfp[dfp['topic']==12].head(50)

,source,video_id,video_url,title,description,channel,comments,views,likes,comments_count,...,country,date,likes_missing,year,engagement,engagement_rate,text,topic,topic_name,topic_repr
127,YouTube,tgQzRaVRb4Q,https://www.youtube.com/watch?v=tgQzRaVRb4Q,BrainPill Remastered,"Hier das alte video in der besten Qualität, di...",crazysoulbrothers,NaN,133.0,1.0,0.0,...,unknown,2010-02-15,False,2010,1.0,0.007519,BrainPill Remastered Hier das alte video in de...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."
382,YouTube,nxtAP2L9gkk,https://www.youtube.com/watch?v=nxtAP2L9gkk,Super Schnelles Schlank Getrank,"My German video, wooooo.",Luke Dowker,haha :) love the accent when they talk in germ...,209.0,3.0,2.0,...,unknown,2010-05-10,False,2010,5.0,0.023923,Super Schnelles Schlank Getrank My German vide...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."
1244,YouTube,ecW4mLg2NFE,https://www.youtube.com/watch?v=ecW4mLg2NFE,Kinetics Live 21 januari,Kinetics spelar live på Kafé de luxe 21 januar...,simontrummor,NaN,468.0,1.0,0.0,...,unknown,2010-12-29,False,2010,1.0,0.002137,Kinetics Live 21 januari Kinetics spelar live ...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."
1253,YouTube,XxoF8UzdIDU,https://www.youtube.com/watch?v=XxoF8UzdIDU,Soxhletextraktion einer Pflanze.mpg,"So stellt man hochwirksame Medizin her, einfac...",Armin M ü l l e r,wielange braucht man dafür<br>? || bringt nich...,1816.0,5.0,6.0,...,unknown,2011-01-02,False,2011,11.0,0.006057,Soxhletextraktion einer Pflanze.mpg So stellt ...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."
1621,YouTube,7uHieI1qqbk,https://www.youtube.com/watch?v=7uHieI1qqbk,Pharmeo | Anti Aging (Hautalterung),http://www.pharmeo.de/advanced_search_result.p...,Pharmeo,NaN,1082.0,2.0,0.0,...,unknown,2011-04-28,False,2011,2.0,0.001848,Pharmeo | Anti Aging (Hautalterung) http://www...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."
1668,YouTube,tSuTCosFJkM,https://www.youtube.com/watch?v=tSuTCosFJkM,Anti-Aging (thuis-)Behandeling,Neways - de producent van eerlijke producten (...,gezondereweg,NaN,275.0,0.0,0.0,...,unknown,2011-05-12,True,2011,0.0,0.000000,Anti-Aging (thuis-)Behandeling Neways - de pro...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."
1733,YouTube,7JwsLYS7a84,https://www.youtube.com/watch?v=7JwsLYS7a84,Anti-Aging bei Fenninger Kosmetik in Salzburg,http://www.fenninger-kosmetik.at Herzlich Will...,WebstyleGmbH,NaN,532.0,1.0,0.0,...,unknown,2011-06-03,False,2011,1.0,0.001880,Anti-Aging bei Fenninger Kosmetik in Salzburg ...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."
1774,YouTube,RDU8ff_2O-s,https://www.youtube.com/watch?v=RDU8ff_2O-s,www.boturine-x.com - Kosmetik Boturine-X Anti-...,http://www.boturine-x.com : Einfache und schme...,boturine,NaN,1207.0,3.0,0.0,...,unknown,2011-06-19,False,2011,3.0,0.002486,www.boturine-x.com - Kosmetik Boturine-X Anti-...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."
1849,YouTube,gGEcCeuUfsE,https://www.youtube.com/watch?v=gGEcCeuUfsE,ANTI AGING - actiMare natural cosmetics - Meer...,mehr infos: www.actimare.de Onlineshop: www.ac...,actiMare,NaN,405.0,2.0,0.0,...,de,2011-07-14,False,2011,2.0,0.004938,ANTI AGING - actiMare natural cosmetics - Meer...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."
1876,YouTube,in7kNoJ_8QA,https://www.youtube.com/watch?v=in7kNoJ_8QA,#27 Ellfiore Testimonial - Anti Aging Pflege,(Werbung) Yuhu - im Jahr 2011 darf ich das Ges...,ElischebaTV,NaN,760.0,7.0,0.0,...,unknown,2011-07-23,False,2011,7.0,0.009211,#27 Ellfiore Testimonial - Anti Aging Pflege (...,12,12_die_und_der_für,"['die', 'und', 'der', 'für', 'ist', 'das', 'si..."


We were right, Let's delete them!

In [529]:
dfp = dfp[(dfp['topic']!=11) & (dfp['topic']!=12)]

In [537]:
tpp = tpp[(tpp['Topic']!=11) & (tpp['Topic']!=12

In [616]:
dfp.to_csv("new_paraphrase_dataset.csv", index=False)
tpp.to_csv("new_paraphrase_topics.csv", index=False)

Topic number 3 is a little suspicious since some of the vidoes are not related. Some vidoes contain the rejuvination of the different things than health (eg. yard, park etc.). Keep or Delete? Some videos contain mental rejuvination musics. 

Other topics seems like more or less related